wen-restart/src/wen_restart.rs

```diff
@@ -737,7 +754,120 @@ pub(crate) fn aggregate_restart_heaviest_fork(
+            stat_printed_at = Instant::now();
+        }
+    }
+}
+
+pub(crate) fn repair_heaviest_fork(
+    my_heaviest_fork_slot: Slot,
+    heaviest_slot: Slot,
+    exit: Arc<AtomicBool>,
+    blockstore: Arc<Blockstore>,
+    wen_restart_repair_slots: Arc<RwLock<Vec<Slot>>>,
+) -> Result<()> {
+    loop {
+        if exit.load(Ordering::Relaxed) {
+            return Err(WenRestartError::Exiting.into());
+        }
+        let to_repair = if blockstore.meta(heaviest_slot).is_ok_and(|x| x.is_some()) {
+            AncestorIterator::new_inclusive(heaviest_slot, &blockstore)
+                .take_while(|slot| *slot > my_heaviest_fork_slot)
+                .filter(|slot| !blockstore.is_full(*slot))
+                .collect()
+        } else {
+            vec![heaviest_slot]
+        };
+        info!("wen_restart repair slots: {:?}", to_repair);
+        if to_repair.is_empty() {
+            return Ok(()); // All blocks are full
+        }
+        *wen_restart_repair_slots.write().unwrap() = to_repair;
+        sleep(Duration::from_millis(GOSSIP_SLEEP_MILLIS));
+    }
+}
+
+pub(crate) fn verify_coordinator_heaviest_fork(
+    my_heaviest_fork_slot: Slot,
+    coordinator_heaviest_slot: Slot,
+    coordinator_heaviest_hash: &Hash,
+    bank_forks: Arc<RwLock<BankForks>>,
+    blockstore: Arc<Blockstore>,
+    exit: Arc<AtomicBool>,
+    wen_restart_repair_slots: Arc<RwLock<Vec<Slot>>>,
+) -> Result<()> {
+    repair_heaviest_fork(
+        my_heaviest_fork_slot,
+        coordinator_heaviest_slot,
+        exit.clone(),
+        blockstore.clone(),
+        wen_restart_repair_slots.clone(),
+    )?;
+    let root_bank;
+    {
+        root_bank = bank_forks.read().unwrap().root_bank();
+    }
+    let root_slot = root_bank.slot();
+    let mut slots: Vec<Slot> =
+        AncestorIterator::new_inclusive(coordinator_heaviest_slot, &blockstore)
+            .take_while(|slot| slot >= &root_slot)
+            .collect();
+    slots.sort();
+    if !slots.contains(&root_slot) {
+        return Err(WenRestartError::HeaviestForkOnLeaderOnDifferentFork(
+            coordinator_heaviest_slot,
+            root_slot,
+        )
+        .into());
+    }
+    if coordinator_heaviest_slot > my_heaviest_fork_slot && !slots.contains(&my_heaviest_fork_slot)
+    {
+        return Err(WenRestartError::HeaviestForkOnLeaderOnDifferentFork(
+            coordinator_heaviest_slot,
+            my_heaviest_fork_slot,
+        )
+        .into());
+    }
+    if coordinator_heaviest_slot < my_heaviest_fork_slot
+        && !AncestorIterator::new(my_heaviest_fork_slot, &blockstore)
+            .any(|slot| slot == coordinator_heaviest_slot)
+    {
+        return Err(WenRestartError::HeaviestForkOnLeaderOnDifferentFork(
+            coordinator_heaviest_slot,
+            my_heaviest_fork_slot,
+        )
+        .into());
+    }
+    let my_bankhash = if !slots.is_empty() {
+        find_bankhash_of_heaviest_fork(
```

Comment:
what happens here if the `coordinator_heaviest_slot` is accurate, our heaviest slot is accurate, but we've repaired blocks with the wrong ancestors, will `find_bankhash_of_heaviest_fork` just error out on replay with a dead slot?

Reply:
`find_bankhash_of_heaviest_fork` blindly follows the given slots to find the bankhash. If the given slots form a valid ancestor chain (checking it's frozen), a bankhash will be returned, then we can find out the bankhash is actually different.

Reply:
ok yeah looks like it will just error out in `process_single_slot` with a replay error if the ancestor chain is wrong